src/radical/pilot/pmgr/launching/base.py

```diff
@@ -25,34 +123,1105 @@ def __init__(self, cfg, session):
+                    self._log.warn('invalid pilot launcher name: %s', lname)
+                    continue
+
+                launcher = self._launchers[lname]
+                try:
+                    launcher.cancel_pilots([pid])
+                except:
+                    self._log.exception('pilot cancel failed for %s' % pid)
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _kill_pilots(self, pids=None):
+        '''
+        Forcefully kill a set of pilots.  For pilots which have just recently be
+        cancelled, we will wait a certain amount of time to give them a chance
+        to termimate on their own (which allows to flush profiles and logfiles,
+        etc).  After that delay, we'll make sure they get killed.
+        '''
+
+        self._log.debug('killing pilots: %s', pids)
+
+        if not self._pilots:
+            # nothing to do
+            return
+
+        if not pids:
+            pids = self._pilots.keys()
+
+        # find the most recent cancellation request
+        with self._pilots_lock:
+            self._log.debug('killing pilots: %s',
+                              [p['pilot'].get('cancel_requested', 0)
+                               for p in list(self._pilots.values())])
+            last_cancel = max([p['pilot'].get('cancel_requested', 0)
+                               for p in list(self._pilots.values())])
+
+        self._log.debug('killing pilots: last cancel: %s', last_cancel)
+
+        # we wait for up to PILOT_CANCEL_DELAY for a pilt
+        while time.time() < (last_cancel + PILOT_CANCEL_DELAY):
+
+            self._log.debug('killing pilots: check %s < %s + %s',
+                    time.time(), last_cancel, PILOT_CANCEL_DELAY)
+
+            alive_pids = list()
+            for pid in pids:
+
+                if pid not in self._pilots:
+                    self._log.error('unknown: %s', pid)
+                    raise ValueError('unknown pilot %s' % pid)
+
+                pilot = self._pilots[pid]['pilot']
+                if pilot['state'] not in rps.FINAL:
+                    self._log.debug('killing pilots: alive %s', pid)
+                    alive_pids.append(pid)
+                else:
+                    self._log.debug('killing pilots: dead  %s', pid)
+
+            pids = alive_pids
+            if not alive_pids:
+                # nothing to do anymore
+                return
+
+            # avoid busy poll)
+            time.sleep(1)
+
+        # we don't want the watcher checking for these pilot anymore
+        with self._check_lock:
+            for pid in pids:
+                if pid in self._checking:
+                    self._checking.remove(pid)
+
+
+        self._log.debug('killing pilots: kill! %s', pids)
+        try:
+            with self._pilots_lock:
+
+                pilots = list()
+                for pid in pids:
+
+                    pilot = self._pilots['uid']['pilot']
```

Comment:
`pilot = self._pilots[pid]['pilot']`

btw, I cannot figure out how `self._pilots` is filled out with pilots, did I miss it or it is actually missed?

Reply:
This is fixed now also - thanks for catching!